File: 105.Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        return buildTreeHelper(0,0,inorder.length-1, preorder,inorder);
    }
    
    private TreeNode buildTreeHelper(int preStart, int inStart, int inEnd, int[] preorder, int[] inorder) {
        // #end condition:
        if (preStart >= preorder.length || inStart > inEnd) return null;
        TreeNode root = new TreeNode(preorder[preStart]);
        int inIndex = 0;
        //for loop 找inoder index for current root
        for (int i=inStart; i<=inEnd; i++) {
            if (inorder[i] == preorder[preStart]){
                inIndex = i;
                break;
            }
        }
        // left start = preStart + 1
        root.left = buildTreeHelper(preStart+1, inStart, inIndex-1, preorder, inorder);
        // right start = preStart + number of left nodes = preStart + (inIndex - inStart)+1
        root.right = buildTreeHelper(preStart+inIndex-inStart+1, inIndex+1, inEnd, preorder, inorder);
        return root;
    }
}

File: 105.ConstructBinaryTreefromPreorderandInorderTraversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        res = self.helper(preorder, inorder, 0, 0, len(inorder)-1)
        return res

    def helper(self, preorder, inorder, preStart, inStart, inEnd):
        # end condition
        if preStart >= len(preorder) or inStart > inEnd:
            return None

        rootVal = preorder[preStart]
        curr = TreeNode(rootVal)
        inIndex = 0
        for i in range(inStart, inEnd+1):
            if inorder[i] == rootVal:
                inIndex = i
                break

        curr.left = self.helper(preorder, inorder, preStart+1, inStart, inIndex-1)
        # right index = preorder + num_leftNodes = preorder + (inIndex - inStart)
        curr.right = self.helper(preorder, inorder, preStart+(inIndex-inStart+1), inIndex+1, inEnd)
        return curr